clang/tools/driver/cc1_main.cpp

```diff
@@ -678,10 +680,32 @@ void CompileJobCache::finishComputedResult(CompilerInstance &Clang,
                                 &Clang.getInvocation().getDiagnosticOpts()),
       /*ShouldOwnClient=*/true);
 
+  // Check if we encounter any source that would generate non-deterministic
+  // outputs.
+  if (Clang.hasPreprocessor() && Clang.isSourceNondeterministic()) {
+    DiagnosticsEngine &Diags = Clang.getDiagnostics();
+    switch (Clang.getFrontendOpts().ProgramAction) {
+      case frontend::GenerateModule:
+      case frontend::GenerateModuleInterface:
+      case frontend::GenerateHeaderModule:
+      case frontend::GeneratePCH:
+        // It is an error when generating modules/pch with cache enabled but
+        // generates non-deterministic output.
+        Diags.Report(diag::err_caching_output_non_deterministic);
+        return false;
+      default:
+        Diags.Report(diag::warn_clang_cache_disabled_caching)
+            << "output can be non-deterministic. Please use -Wdate-time to "
```

Comment:
I think we should attempt to provide a more specific diagnostic - either force -Wdate-time on when caching, or else keep track of the first use of one of these macros and the source location so we can diagnose it here precisely.

Reply:
Nah, this warning is about we are skipping caching and I don't know about emitting a different diagnostics at the same place where other warning can be emitted. I can ask `clang-cache` launch to add `-Wdate-time` by default, or you think it is better to emit caching skipped warning in Preprocessor?